File: python/crypto_api/math_lib/multiplicative_group.py
import random

from crypto_api.math_lib.misc import phi, ext_factors


class MultiplicativeGroup:

    def __init__(self, modulo: int) -> None:
        self.__m = modulo
        self.__order = phi(ext_factors(modulo))

    def mul_mod(self, x: int, y: int) -> int:
        return ((x % self.__m) * (y % self.__m)) % self.__m

    def pow_mod(self, base: int, power: int) -> int:
        result = 1
        for i in range(power):
            result = self.mul_mod(result, base)
        return result

    def mul_invert_mod(self, x) -> int:
        return x ** (self.__order - 1) % self.__m

    def random_element(self) -> int:
        return random.randint(0, self.__m - 1)

    @property
    def order(self) -> int:
        return self.__order

    @property
    def modulo(self) -> int:
        return self.__m

File: java/cryptoapi/math_lib/MultiplicativeGroup.java
package cryptoapi.math_lib;

import java.math.BigInteger;
import java.security.SecureRandom;

public class MultiplicativeGroup {
    public final BigInteger m;
    public final BigInteger order;
    private final SecureRandom random = new SecureRandom();

    public MultiplicativeGroup(BigInteger modulo) {
        this.m = modulo;
        this.order = MathMisc.phi(MathMisc.summariseFactors(MathMisc.factorize(modulo)));
    }

    public BigInteger mulMod(BigInteger x, BigInteger y) {
        return x.multiply(y).mod(m);
    }

    public BigInteger powMod(BigInteger base, BigInteger power) {
        return base.modPow(power, this.m);
    }

    public BigInteger mulInvertMod(BigInteger x) {
        return x.modPow(order.subtract(BigInteger.ONE), m);
    }

    public BigInteger wrapping(BigInteger x) {
        return x.mod(m).add(m).mod(m);
    }

    public BigInteger randomElement() {
        return new BigInteger(m.bitLength(), random).mod(m);
    }
}
